### ISSUE TITLE
Add `replace: bool` argument to `random.categorical` to sample without replacement using Gumbel-top-k trick

### ISSUE THREAD

**Issue by carlosgmartin** [jax-ml/jax#25617]
[`random.categorical`](https://jax.readthedocs.io/en/latest/_autosummary/jax.random.categorical.html) uses the Gumbel-max trick to sample with replacement.

An extension of the Gumbel-max trick allows sampling _without_ replacement. As noted in [this paper](https://arxiv.org/abs/1903.06059):

> The well-known Gumbel-Max trick for sampling from a categorical distribution can be extended to sample k elements without replacement.

> The Gumbel-Max trick [...] allows sampling from the categorical distribution, simply by _perturbing_ the log-probability for each category by adding _independent_ Gumbel distributed noise and returning the category with maximum _perturbed log-probability_. [...] However, there is more: as was noted (in a blog post) by Vieira (2014), taking the **top k largest perturbed log-probabilities** (instead of the maximum, or _top 1_) yields a sample of size k from the categorical distribution _without replacement_. We refer to this extension of the Gumbel-Max trick as the _Gumbel-Top-k_ trick.

**Proposal:**

Add a `replace: bool` argument to `random.categorical`. `replace=False` should use the Gumbel-Top-k trick (replacing [`argmax`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.argmax.html) with [`top_k`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.top_k.html)) to sample without replacement.

**Alternatives:**

[`random.choice`](https://jax.readthedocs.io/en/latest/_autosummary/jax.random.choice.html) can sample without replacement, but uses a less efficient [`cumsum`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.cumsum.html)-based method, and operates on probabilities rather than logits/log-probabilities. Related:
- https://github.com/jax-ml/jax/issues/22682
- https://github.com/jax-ml/jax/issues/25498

**Comment by jakevdp** [comment]
Sounds reasonable – is this something you'd like to contribute?

**Comment by carlosgmartin** [comment]
Yes, I can create a PR for this.

**Comment by carlosgmartin** [comment]
@jakevdp After thinking about it more, since the semantics of shapes and axes would differ from those of [sampling with replacement](https://jax.readthedocs.io/en/latest/_autosummary/jax.random.categorical.html), I think it might be better to create a separate function. What would be a good name for it? Here are a few candidates:
- `categorical_wor`
- `categorical_without_replacement`
- `categorical_no_replace`
- `categorical_no_replacement`
- `sample_without_replacement`

Here's an example implementation:

```python3
from jax import lax, nn, numpy as jnp, random


def categorical_wor(key, logits, k, axis=-1, batch_shape=()):

    logits = jnp.broadcast_to(logits, batch_shape + logits.shape)
    if axis >= 0:
        axis += len(batch_shape)

    noise = random.gumbel(key, logits.shape, logits.dtype)
    x = logits + noise

    # return x.argmax(axis)  # sampling with replacement

    x = jnp.moveaxis(x, axis, -1)
    _, outcomes = lax.top_k(x, k)
    outcomes = jnp.moveaxis(outcomes, -1, axis)

    return outcomes


def get_freqs(x, length):
    counts = jnp.bincount(x.flatten(), length=length)
    return counts / counts.sum()


def main():
    key = random.key(0)

    key, subkey = random.split(key)
    logits = jnp.arange(4, dtype=float)
    print(f"probs: {nn.softmax(logits)}")

    axis = -1

    for k in range(logits.shape[axis] + 1):
        outcomes = categorical_wor(key, logits, k, axis, (20, 30, 40))
        print(f"{k=} freqs: {get_freqs(outcomes, logits.shape[axis])}")


if __name__ == "__main__":
    main()
```

```
probs: [0.0320586  0.08714432 0.23688284 0.6439143 ]
k=0 freqs: [nan nan nan nan]
k=1 freqs: [0.03220833 0.08495833 0.23579167 0.6470417 ]
k=2 freqs: [0.05175    0.13675    0.34770834 0.46379167]
k=3 freqs: [0.10051389 0.25186113 0.31643057 0.33119443]
k=4 freqs: [0.25 0.25 0.25 0.25]
```

(As expected, the sample frequencies become equal by the time we get to `k == logits.shape[axis]`.)

**Comment by jakevdp** [comment]
Can you say more about why this needs its own function?

**Comment by carlosgmartin** [comment]
The output shape of sampling with replacement [is](https://jax.readthedocs.io/en/latest/_autosummary/jax.random.categorical.html) `delete(logits.shape, axis)`.

The output shape of sampling without replacement would be `insert(delete(logits.shape, axis), axis2, k)`, where `k` is the number of draws in the sequential sampling-without-replacement process and `axis2` is some axis (possibly `axis`).

**Comment by carlosgmartin** [comment]
@jakevdp Thoughts on the above?

**Comment by jakevdp** [comment]
> The output shape of sampling with replacement [is](https://jax.readthedocs.io/en/latest/_autosummary/jax.random.categorical.html) `delete(logits.shape, axis)`.

This is only true if `shape` is not specified. We could keep the same semantics for sampling without replacement, I think: the default would be sampling a single value without replacement, and if `shape` were provided we would sample more than one value without replacement.

Would that work?

**Comment by jakevdp** [comment]
Why specify `batch_shape` and `samples`, rather than just specifying a single `shape` as in `random.categorical`?

**Comment by jakevdp** [comment]
Also, how do you control which axis corresponds to logits for a single distribution? Are we just assuming it's the last axis?

**Comment by jakevdp** [comment]
I guess overall I'm just confused as to why sampling with and without replacement need such different APIs.

**Comment by carlosgmartin** [comment]
@jakevdp Before answering those questions, could you clarify the input and output semantics you have in mind, especially the meaning of the input and output shape and axes in the case of sampling without replacement? (See my questions at https://github.com/jax-ml/jax/issues/25617#issuecomment-2683333976.)

**Comment by jakevdp** [comment]
I'm basing my assumptions here on the API of `https://docs.jax.dev/en/latest/_autosummary/jax.random.categorical.html`

If you provide no shape, you get a single value with shape `np.delete(logits.shape, axis)`

If you provide a shape, it must broadcast with the `np.delete(logits.shape, axis)`, and you are sampling multiple values per distribution (with replacement).

My question is: why couldn't sampling without replacement have the same semantics as sampling with replacement?

**Comment by jakevdp** [comment]
More concretely, for example, here is how you might sample five values with replacement:
```python
In [3]: p = jnp.arange(1, 10)

In [4]: logits = jnp.log(logits / logits.sum())

In [5]: jax.random.categorical(jax.random.key(0), logits, axis=0, shape=(5,))
Out[5]: Array([1, 2, 0, 0, 1], dtype=int32)
```
It seems like you should be able to sample five values without replacement by doing something like this:
```python
jax.random.categorical(jax.random.key(0), logits, axis=0, shape=(5,), with_replacement=False)
```

**Comment by carlosgmartin** [comment]
Does this match the exact shape and axis semantics you have in mind? In particular, note the line `indices = jnp.moveaxis(indices, axis, 0)`. I'm not sure whether you always want to push the `top_k` results to the front.

```python3
def tuple_delete(tup, index):
    return tup[:index] + tup[index:][1:]


def top_k(operand, k, axis):
    # jax/issues/25867
    operand = jnp.moveaxis(operand, axis, -1)
    values, indices = lax.top_k(operand, k)
    values = jnp.moveaxis(values, -1, axis)
    indices = jnp.moveaxis(indices, -1, axis)
    return values, indices


def new_categorical(key, logits, axis=-1, shape=None, replace=True):
    if replace:
        return random.categorical(key, logits, axis, shape)

    if shape is None:
        shape = tuple_delete(logits.shape, axis)

    logits += random.gumbel(key, logits.shape)

    batch_shape = shape[: len(shape) - logits.ndim + 1]
    batch_size = prod(batch_shape)

    _, indices = top_k(logits, batch_size, axis)

    indices = jnp.moveaxis(indices, axis, 0)

    indices = indices.reshape(shape)

    return indices


def main():
    key = random.key(0)

    logits = jnp.zeros((20, 2, 5))
    output = new_categorical(key, logits, 0, (7, 2, 5), False)
    assert output.shape == (7, 2, 5)

    logits = jnp.zeros((2, 20, 5))
    output = new_categorical(key, logits, 1, (7, 2, 5), False)
    assert output.shape == (7, 2, 5)

    logits = jnp.zeros((2, 5, 20))
    output = new_categorical(key, logits, 2, (7, 2, 5), False)
    assert output.shape == (7, 2, 5)

    logits = jnp.zeros((20, 2, 5))
    output = new_categorical(key, logits, 0, None, False)
    assert output.shape == (2, 5)

    logits = jnp.zeros((2, 20, 5))
    output = new_categorical(key, logits, 1, None, False)
    assert output.shape == (2, 5)

    logits = jnp.zeros((2, 5, 20))
    output = new_categorical(key, logits, 2, None, False)
    assert output.shape == (2, 5)
```

**Comment by jakevdp** [comment]
Yeah, that looks reasonable

**Comment by carlosgmartin** [comment]
@jakevdp I've opened a PR: https://github.com/jax-ml/jax/pull/26980.
